### PR TITLE
feat/#76 implement roll back to

### PR DESCRIPTION
**Pull request by danielvo11**
- allow user to config rollbackFunction in AloxideActionHandler
- reset index state to block number when rollback

Require https://github.com/lecle/demux-js/pull/2 to be merged first

### CODE REVIEW

File: packages/demux/src/AloxideActionHandler.ts [hunk 1]
```diff
@@ -78,6 +80,20 @@ export class AloxideActionHandler extends AbstractActionHandler {
       .then<void>(() => {});
   }
 
+  private async rollbackIndexState(blockNum: number): Promise<void> {
```
Comment (tiendq) [inline packages/demux/src/AloxideActionHandler.ts:83]: Why `blockNum` but not `blockNumber` same as other places?

File: packages/demux/src/AloxideActionHandler.ts [hunk 2]
```diff
@@ -78,6 +80,20 @@ export class AloxideActionHandler extends AbstractActionHandler {
...
+    this.indexStateModel.blockHash = '';
+
+    return this.dataAdapter
+      .update(this.getIndexStateModelName(), this.indexStateModel, {
+        entity: this.indexStateEntity,
+      })
+      .catch(err => {
+        this.log.error('---- demux updateIndexState error:', err);
+      })
+      .then<void>(() => {});
```
Comment (tiendq) [inline packages/demux/src/AloxideActionHandler.ts:94]: What's its purpose?

File: packages/demux/src/createWatcher.ts [hunk 3]
```diff
@@ -25,6 +25,7 @@ export interface CreateWatcherConfig {
   actionHandler?: ActionHandler;
   actionHandlerOptions?: AloxideActionHandlerOptions;
   actionWatcherOptions?: ActionWatcherOptions;
+  rollbackFunction?: (blockNum: number) => Promise<void>;
```
Comment (tiendq) [inline packages/demux/src/createWatcher.ts:28]: Naming should be consistent :)

File: packages/demux/src/AloxideActionHandler.ts (hunk 1 above)
Comment (danielAlvess) [reply packages/demux/src/AloxideActionHandler.ts]: updated

File: packages/demux/src/createWatcher.ts (hunk 3 above)
Comment (danielAlvess) [reply packages/demux/src/createWatcher.ts]: updated

File: packages/demux/src/AloxideActionHandler.ts (hunk 2 above)
Comment (danielAlvess) [reply packages/demux/src/AloxideActionHandler.ts]: it update last processed block in db to roll back block number. After rollback, demux can continue reading from that block num.

Comment (tiendq) [reply packages/demux/src/AloxideActionHandler.ts]: I was only curious about this dumb function :)

` .then<void>(() => {});`

Comment (danielAlvess) [reply packages/demux/src/AloxideActionHandler.ts]: updated

**Comment by manh-vv** [comment]
Let's implement this feature without introduce breaking change in the `demux-js`.

```typescript
  protected async rollbackTo(blockNumber: number): Promise<void> {
    this.log.debug('-- roll back to block number:', blockNumber);
    // - [ ] TODO alter current index state
    // - [ ] TODO revert all change from blockNumber to current blockNumber
    // Aloxide Updater will create a snapshot of data when applying a last irreversible block
    // doing so we can quickly roll back to the last irreversible block state
    // This is only applied to Aloxide data
  }

  public async handleBlock(nextBlock: NextBlock, isReplay: boolean): Promise<number | null> {
    try {
      return super.handleBlock(nextBlock, isReplay);
    } catch (e) {
      if (e instanceof MismatchedBlockHashError) {
        // roll back to the last irreversible blocks
        // - [ ] TODO explain why it must be the last irreversible block
        // - [ ]  what hapen if the rollbackTo took too long to execute
        await this.rollbackTo(this.lastIrreversibleBlockNumber);
        return this.lastIrreversibleBlockNumber + 1;
      }

      throw e;
    }
  }
```

cc @danielAlvess

**Comment by danielAlvess** [comment]
@manh-vv please help me review again!.

File: packages/demux/src/AloxideActionHandler.ts [hunk 4]
```diff
@@ -143,6 +168,52 @@ export class AloxideActionHandler extends AbstractActionHandler {
...
   }
 
+  public async handleBlock(nextBlock: NextBlock, isReplay: boolean): Promise<number | null> {
+    try {
+      return await super.handleBlock(nextBlock, isReplay);
+    } catch (e) {
+      if (e instanceof MismatchedBlockHashError) {
+        if (this.lastProcessedBlockHash) {
+          /* ActionReader store processed block info in ephemeral memory,
+            in case of they lost it, they don't can detect a fork and return new block to ActionHandler
```
Comment (manh-vv) [inline packages/demux/src/AloxideActionHandler.ts:178]: Grammar error, please fix ['they', [don't can]]

File: packages/demux/src/AloxideActionHandler.ts [hunk 5]
```diff
@@ -30,7 +30,7 @@ export class AloxideActionHandler extends AbstractActionHandler {
 
   constructor(
     protected bcName: string,
-    protected dataAdapter: DataAdapter<any, any>,
+    protected dataAdapter: AloxideDataManager,
```
Comment (manh-vv) [inline packages/demux/src/AloxideActionHandler.ts:33]: I saw that there are two more functions that you added to `AloxideDataManager`. Why don't you just add them to `DataAdapter` and remove this change?

File: packages/demux/src/AloxideActionHandler.ts [hunk 6]
```diff
@@ -143,6 +168,52 @@ export class AloxideActionHandler extends AbstractActionHandler {
...
+          await this.handleWithState(handleWithArgs);
+          return null;
+        }
+      }
+
+      throw e;
+    }
+  }
+
+  private async handleRollbackToLastIreversibleBlock() {
```
Comment (manh-vv) [inline packages/demux/src/AloxideActionHandler.ts:203]: why does this need to be a function alone?

File: packages/demux/src/AloxideActionHandler.ts [hunk 7]
```diff
@@ -153,6 +224,7 @@ export class AloxideActionHandler extends AbstractActionHandler {
   }
 
   protected async rollbackTo(blockNumber: number): Promise<void> {
-    this.log.debug('-- roll back to block number:', blockNumber);
+    await this.rollbackIndexState(blockNumber);
+    await this.dataAdapter.revert(blockNumber);
```
Comment (manh-vv) [inline packages/demux/src/AloxideActionHandler.ts:228]: I think if `rollbackIndexState` is executed then we don't need `dataAdapter.revert`.

File: packages/demux/src/AloxideDataManager.ts [hunk 8]
```diff
@@ -68,6 +68,22 @@ export class AloxideDataManager implements DataAdapter {
     return this.getDataProvider(name).delete(id, meta);
   }
 
+  async revert(blockNumber: number): Promise<any> {
```
Comment (manh-vv) [inline packages/demux/src/AloxideDataManager.ts:71]: Please add a document to explain this function.

File: packages/demux/src/AloxideDataManager.ts [hunk 9]
```diff
@@ -68,6 +68,22 @@ export class AloxideDataManager implements DataAdapter {
...
+    for (const [name, adapter] of this.dataProviderMap.entries()) {
+      if (adapter.revert) {
+        await adapter.revert(blockNumber);
+      }
+    }
+  }
+
+  async deleteAuditHistory(blockNumber: number): Promise<any> {
```
Comment (manh-vv) [inline packages/demux/src/AloxideDataManager.ts:79]: Please add a document to explain this function.

File: packages/demux/src/DataAdapter.ts [hunk 10]
```diff
@@ -60,4 +60,18 @@ export interface DataAdapter<ID = any, D = any> {
    * @param meta metadata (to be updated)
    */
   delete(name: modelName, id: ID, meta: DMeta): Promise<boolean>;
+
+  /**
+   * revert data to state of blockNumber in case of fork
+   * @param blockNumber the point to revert to
+   * revert data will be from blockNumber + 1 to current processed block
+   */
+  revert?(blockNumber): Promise<any>;
```
Comment (manh-vv) [inline packages/demux/src/DataAdapter.ts:69]: I don't think revert and deleteAuditHistory should exist in the DataAdapter layer.

File: packages/demux/src/DataProvider.ts [hunk 11]
```diff
@@ -60,4 +60,8 @@ export interface DataProvider<ID = any, D = any> {
    * @param meta metadata (to be updated)
    */
   delete(id: ID, meta: DMeta): Promise<boolean>;
+
+  revert?(blockNumber: number): Promise<boolean>;
```
Comment (manh-vv) [inline packages/demux/src/DataProvider.ts:64]: I don't think revert and deleteAuditHistory should exist in the DataProvider.

**Review by manh-vv — CHANGES_REQUESTED** [review]
Please have a direct talk if you have any questions about my comments.

File: packages/demux/src/AloxideActionHandler.ts (hunk 5 above)
Comment (danielAlvess) [reply packages/demux/src/AloxideActionHandler.ts:33]: With `DataAdapter` we don't know which entityNames to revert.

File: packages/demux/src/AloxideActionHandler.ts (hunk 7 above)
Comment (danielAlvess) [reply packages/demux/src/AloxideActionHandler.ts:228]: `rollbackIndexState` is roll back data of index state, it update lastProcessedBlockNumber to rollback blockNumber,

`dataAdapter.revert` is revert data of entity such as `poll`, `option`, `vote`

File: packages/demux/src/AloxideActionHandler.ts (hunk 6 above)
Comment (danielAlvess) [reply packages/demux/src/AloxideActionHandler.ts:203]: this function create debug log about rollback, call function to roll back data and roll back deferred effects.

File: packages/demux/src/DataAdapter.ts (hunk 10 above)
Comment (danielAlvess) [reply packages/demux/src/DataAdapter.ts:69]: So could you recommend how to implement revert?

**Review by manh-vv — APPROVED** [review]
LGTM

**Pull request merged**